### PR TITLE
feat: LWW-Element-Set & Tests

### PR DESCRIPTION
**Pull request by joaopereira12**
LWW-Element-Set & OR-Set implemented in the builtins.

### CODE REVIEW

**Comment by joaopereira12** [comment]
I've finished the tests and the implementations. 
I also made the following changes in the existing CRDTs:
- I've fixed the compare method (instead of comparing with '==' it compares the values of each set)
- I've change the ID of the GCounter from string to number (I've read that the ID is a number from 0 to n-1) (https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#G-Counter_(Grow-only_Counter))

File: packages/crdt/src/builtins/GCounter/index.ts [hunk 1]
```diff
@@ -1,11 +1,19 @@
+export interface IGCounter {
+  globalCounter: number;
+  counts: { [nodeKey: number]: number };
+  value(): number;
+  increment(nodeId: number, amount: number): void;
+  compare(peerCounter: IGCounter): boolean;
+  merge(peerCounter: IGCounter): void;
+}
```
Comment (d-roak) [inline packages/crdt/src/builtins/GCounter/index.ts:8]: Why?

File: packages/crdt/src/builtins/GCounter/index.ts [hunk 2]
```diff
@@ -1,11 +1,19 @@
...
 /// GCounter with support for state and op changes
 export class GCounter {
   globalCounter: number;
   // instead of standard incremental id for replicas
   // we map the counter with the node id
-  counts: { [nodeId: string]: number };
```
Comment (d-roak) [inline packages/crdt/src/builtins/GCounter/index.ts:6]: We are mapping nodeId that come from libp2p. They are not numbers

File: packages/crdt/src/builtins/GCounter/index.ts [hunk 3]
```diff
@@ -28,14 +36,19 @@ export class GCounter {
...
+
+  merge(peerCounter: IGCounter): void {
+    let temp: { [nodeKey: number]: number } = Object.assign(
       {},
       this.counts,
       peerCounter.counts,
     );
-    Object.keys(temp).forEach((key) => {
+
+    for (let key in Object.keys(temp)) {
```
Comment (d-roak) [inline packages/crdt/src/builtins/GCounter/index.ts:47]: Why change?

File: packages/crdt/src/builtins/GCounter/index.ts [hunk 4]
```diff
@@ -28,14 +36,19 @@ export class GCounter {
...
       this.counts[key] = Math.max(this.counts[key], peerCounter.counts[key]);
-    });
+
+      this.globalCounter = Object.values(this.counts).reduce((a, b) => a + b, 0);
```
Comment (d-roak) [inline packages/crdt/src/builtins/GCounter/index.ts:50]: This shouldn't be inside the loop

File: packages/crdt/src/builtins/GSet/index.ts [hunk 5]
```diff
@@ -19,10 +19,10 @@ export class GSet<T> {
   }
 
   compare(peerSet: GSet<T>): boolean {
-    return this._set === peerSet.set();
+    return (this._set.size == peerSet.set().size && [...this._set].every(value => peerSet.set().has(value)));
```
Comment (d-roak) [inline packages/crdt/src/builtins/GSet/index.ts:22]: Doesn't Set type has a native every fn or something that does the same?

**Review by d-roak — CHANGES_REQUESTED** [review]
Initial review

File: packages/crdt/src/builtins/LWWElementSet/index.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    lookup(element: T): boolean {
+
+        const addElemTimestmp = this._adds.get(element);
+        const remElemTimestmp = this._removes.get(element);
+
+        if (addElemTimestmp !== undefined &&
+            (remElemTimestmp === undefined || addElemTimestmp > remElemTimestmp)) {
+            return true;
+        }
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWElementSet/index.ts:25]: properly write "Timestamp"

File: packages/crdt/src/builtins/LWWElementSet/index.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        return false;
+    }
+
+    add(element: T, timestamp: number): void {
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWElementSet/index.ts:29]: force timestamp to javascript implementation, don't accept it as an argument (same for other fns)

File: packages/crdt/src/builtins/LWWElementSet/index.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    getAdds(): Map<T, number> {
+        return this._adds;
+    }
+
+    getRemoves(): Map<T, number> {
+        return this._removes;
+    }
+
+    areSetsEqual(set1: Set<T>, set2: Set<T>): boolean {
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWElementSet/index.ts:45]: put this fn outside the class as a helper fn (private) and rename it to `compareSets`

File: packages/crdt/src/builtins/LWWElementSet/index.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        return (this.areSetsEqual(adds, otherAdds) && this.areSetsEqual(rems, otherRems));
+
+    }
+
+    merge(peerSet: LWWElementSet<T>): void {
+
+        for (let element of peerSet._adds.keys()) {
+            if (!this._adds.has(element) || this.compareTimestamp(this._adds.get(element)!, peerSet.getAdds().get(element)!)) {
+                this._adds.set(element, peerSet.getAdds().get(element)!);
+            }
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWElementSet/index.ts:64]: why are u making the `!` assertion? also, why not just blindly add the elements?

File: packages/crdt/src/builtins/LWWElementSet/index.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    private compareTimestamp(thisTime: number, otherTime: number) {
+
+        const timestampComparison = thisTime - otherTime;
+
+        if (timestampComparison === 0) {
+            return this._bias !== Bias.ADD;
+        }
+        return timestampComparison < 0;
+    }
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWElementSet/index.ts:82]: put outside the class

Comment (d-roak) [reply packages/crdt/src/builtins/LWWElementSet/index.ts:82]: and add bias as a parameter

**Review by d-roak — CHANGES_REQUESTED** [review]
Comments on LWWSet

**Comment by joaopereira12** [comment]
This branch also contains some changes in the GSet and contains the GSet and 2PSet tests. I've also added the GCounter and PNCounter to pass the pipeline tests, but I'll add them to their PR with the changes that I made.

**Pull request closed without merging**